core/dbt/config/profile.py

```diff
@@ -91,21 +97,19 @@ def to_profile_info(self, serialize_credentials=False):
         result = {
             'profile_name': self.profile_name,
             'target_name': self.target_name,
-            'config': self.config.to_dict(),
+            'config': self.config,
             'threads': self.threads,
             'credentials': self.credentials,
         }
         if serialize_credentials:
-            result['credentials'] = result['credentials'].to_dict()
+            result['config'] = self.config.to_dict()
+            result['credentials'] = self.credentials.to_dict()
         return result
 
-    def __str__(self):
-        return str(self.to_profile_info())
-
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
         if not (isinstance(other, self.__class__) and
                 isinstance(self, other.__class__)):
-            return False
+            return NotImplemented
```

Comment:
do we want to return `NotImplemented` here, or should this raise an internal exception?

Reply:
I think the canonical correct behavior is returning `NotImplemented` in this situation: https://docs.python.org/3/library/constants.html#NotImplemented
and that devolves into `False` if the types don't match.